### ISSUE TITLE
"https" situation over Apache/Ngynx in proxy reverse configuration

### ISSUE THREAD

**Issue by vigotecnologia** [aspnet/AspNetCore#9355]
Situation:

- The kestrel server is running in localhost over simple "http". (http://127.0.0.1:5000)

- Apache/Ngynx is configured using proxy-reverse to provide an endpoint secure to the end-user, example: https://app.domain.com -> proxy-reverse to -> http://127.0.0.1:5000

Well, lets take a look in "Unauthorized controllers".

When we put the decoration [Authorize] in some controller and try to access this controller without do the login proccess, the aspnet automatically will redirect you to the "login page", THE PROBLEM is that the THIS REDIRECT  is done using "http" unstead "https", obviously because the kestrel server is running in "http".

This wouldn't be a problem if in Apache/Ngynx we ALSO setup the endpoint http(80) making a permanent redirect to https(443), in browsers this is not noticeable **BUT in PWA APP running in iOS this will cause the link opening out of the standalone browser, forcing a new browser instance of safari to be opened** because **http**://app.domain.com has a different scope than **https**://app.domain.com

Another problem is if we do not setup the http(80) permanent redirection, to avoid any type of unsecure request, if we do this, the auto redirection will fail.

The aspnet core runtime uses the end-user address "app.domain.com" in your internally requests (this can be seen in /vsr/log/messages) but doesn't consider the http protocol, it is using the kestrel configuration unstead apache.

Hard to tell if this is a issue to be solved in aspnet core or just about the situation, to solve my problem I am running the kestrel facing directly the internet with certificate.

**Comment by irwiss** [comment]
i'd check if you correctly configured your reverse proxy headers and using UseForwardedHeaders with your proxy whitelisted in it's options

**Comment by vigotecnologia** [comment]
I did this, the proxy-reverse in apache is correctly configured and also the useforwardheaders in app startup.

The question isn´t the proxy-reverse, like I said before, the ASPNET Core is using http in your own redirects even with https in the endpoint user.

**Comment by blowdart** [comment]
Did you check in incoming headers inside the app? If apache is configured correctly to forward the x-forwarded-proto header and the headers middleware then links generated should come out as HTTPS rather than HTTP.

**Comment by vigotecnologia** [comment]
This is the apache virtual-host configuration:

<VirtualHost *:443>
    ServerName app.domain.com

    ProxyPreserveHost On
    ProxyPass / http://127.0.0.1:5000/
    ProxyPassReverse / http://127.0.0.1:5000/

    DocumentRoot "/home/a_sites/app"
    ErrorLog /var/log/httpd/app_error.log
    CustomLog /var/log/httpd/app_access.log common

    SSLCertificateFile /certificate/fullchain.pem
    SSLCertificateKeyFile /certificate/privkey.pem
    Include /certificate/options-ssl-apache.conf
</VirtualHost>

**Comment by vigotecnologia** [comment]
And this in the Configure method:

[app.UseForwardedHeaders(new ForwardedHeadersOptions() { ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto });](url)

**Comment by Tratcher** [comment]
See this troubleshooting guide:
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.2

**Comment by Tratcher** [comment]
What's your Asp.Net Core version? We've fixed several bugs here across various releases.

**Comment by vigotecnologia** [comment]
```
C:\Users\roliv>dotnet --info
SDK do .NET Core (refletindo qualquer global.json):
 Version:   2.2.105
 Commit:    7cecb35b92

Ambiente de tempo de execução:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.105\

Host (useful for support):
  Version: 2.2.3
  Commit:  6b8ad509b6

.NET Core SDKs installed:
  1.1.11 [C:\Program Files\dotnet\sdk]
  1.1.13 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.2.105 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

C:\Users\roliv>
```

**Comment by Tratcher** [comment]
And the troubleshooting output from the proxy doc?

**Comment by vigotecnologia** [comment]
Good news,

With the configuration below everything is working fine !!! (thanks for the link @Tratcher)

**Before any middleware:**

`                app.Use((context, next) =>
                {
                    context.Request.Scheme = "https";
                    return next();
                });`

The "local" requests still doing over http but to the end-user everything runs in https, even the runtime redirects.

Now, with an app running in proxy-reverse with apache, WITHOUT a permanent redirection for app.domain.com:80 does not ocurr the fail, the browser gets a **307 internal redirect** and has a behavior as expected.

`Apr 15 16:22:26 saopaulo-2 dotnet-vigoweb: No XML encryptor configured. Key {d6d0fbe2-df4f-4916-ba61-08915191f895} may be persisted to storage in unencrypted form.
Apr 15 16:22:27 saopaulo-2 dotnet-vigoweb: Hosting environment: Production
Apr 15 16:22:27 saopaulo-2 dotnet-vigoweb: Content root path: /home/a_sites/vigoweb
Apr 15 16:22:27 saopaulo-2 dotnet-vigoweb: Now listening on: http://localhost:5000
Apr 15 16:22:27 saopaulo-2 dotnet-vigoweb: Application started. Press Ctrl+C to shut down.
`

And inside the browser:

`307 Internal Redirect`

Thank you everyone that helped with this situation, long live to AspNet Core. :)

**Comment by Tratcher** [comment]
@vigotecnologia note setting scheme directly is a last resort. It's better to get UseForwardedHeaders working.

**Comment by vigotecnologia** [comment]
Hello @Tratcher, after your advise I decided to play a little with Apache configuration and I discovered what was happening. my fault, sorry :)

If we add the `RequestHeader set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}` in a "global" way, the problem occur, screenshots attached, but **if we put this configuration inside the each VirtualHost** everything works fine, with or without setting the scheme directly on AspNet Configure method.

This proves that my problem had'nt nothing about aspnet core runtime, it was just the Apache configuration.

Thanks a lot for everyone that helped, sorry for the inconvenience.

![Anotação 2019-04-15 201420](https://user-images.githubusercontent.com/29063384/56173623-64bb5600-5fbc-11e9-9511-0185ad858f3c.jpg)
![Anotação 2019-04-15 201423](https://user-images.githubusercontent.com/29063384/56173624-64bb5600-5fbc-11e9-9193-a874550e06bc.jpg)
